wallet-new/test/unit/Test/Spec/Addresses.hs

```diff
@@ -348,6 +395,137 @@ spec = describe "Addresses" $ do
+                            Handlers.listAddresses layer (RequestParams pp)
+                        res1 <- runExceptT $ runHandler' $ do
+                            Handlers.listAddresses layer (RequestParams pp1)
+                        res2 <- runExceptT $ runHandler' $ do
+                            Handlers.listAddresses layer (RequestParams pp2)
+                        res3 <- runExceptT $ runHandler' $ do
+                            Handlers.listAddresses layer (RequestParams pp3)
+                        case (res, res1, res2, res3) of
+                            (Right wr, Right wr1, Right wr2, Right wr3) -> do
+                                length (wrData wr) `shouldBe` expectedTotal
+                                let con = wrData wr1 <> wrData wr2 <> wrData wr3
+                                length con `shouldBe` expectedTotal
+                                S.fromList con `shouldBe` S.fromList (wrData wr)
+                                (addrRoot . V1.unV1 . V1.addrId <$> con)
+                                    `shouldBe` (addrRoot . V1.unV1 . V1.addrId <$> wrData wr)
+                            _        -> fail ("Got " ++ show res)
+
+
+            prop "yields the correct ordered resutls" $ withMaxSuccess 20 $ do
+                monadicIO $ do
+                  forM_ [(4,8), (6,6), (5,7)] $ \(acc,adr) ->
+                    withAddressesFixtures acc adr $ \_ layer _ _ -> do
+                        forM_ [2..10] $ \k -> do
+                            let indexes = [1..k]
+                            let (expectedTotal :: Int) = expectedNumber acc adr
+                            let pagesParams = map (\i -> PaginationParams (Page i) (PerPage (quot expectedTotal k + 1)))
+                                              indexes
+                            let pp = PaginationParams (Page 1) (PerPage 50)
+                            res <- runExceptT $ runHandler' $ do
+                                Handlers.listAddresses layer (RequestParams pp)
+                            eiResultsArray <- forM pagesParams $ \ppi -> runExceptT $ runHandler' $ do
+                                Handlers.listAddresses layer (RequestParams ppi)
+                            let resultsArray = sequence eiResultsArray
+                            case (res, resultsArray) of
+                                (Right wr, Right wrList) -> do
+                                    let con = mconcat $ map wrData wrList
+                                    length (wrData wr) `shouldBe` expectedTotal
+                                    length con `shouldBe` expectedTotal
+                                    S.fromList con `shouldBe` S.fromList (wrData wr)
+                                    (addrRoot . V1.unV1 . V1.addrId <$> con)
+                                            `shouldBe` (addrRoot . V1.unV1 . V1.addrId <$> wrData wr)
+                                _        -> fail ("Got " ++ show res)
```

Comment:
This is a minor thing, but these tests feels very imperative rather than declarative. I don't know what we can do about it, but I suspect we might come up with some combinators which factor away duplication and mechanical reshuffling of the data and leave the logic of the tests exposes. I guess what we aim is something like:

```
myTest = do
    -- Initialisation
    -- Execution
    -- Checking of properties
```

Which is exactly what your tests are also doing, but each of these steps takes a lot of lines of code and is hard to track down the spirit of the test 😉